README.rst
.. _Cyclus: http://www.fuelcycle.org/
.. _Eigen: http://eigen.tuxfamily.org/index.php?title=Main_Page

*****************************
Welcome New Bright-lite User!
*****************************
Bright-lite is a collection of modules for the Cyclus_ Fuel Cycle Simulator 
that allow for medium fidelity reactor modeling. There are currently 
three modules in the Bright-lite suite. 

- Bright-lite Reactor Facility
- Bright-lite Fuel Fabrication Facility
- Bright-lite Reprocessing Facility

**Bright-lite Reactor Facility**: A reactor modeling software that uses burnup, criticality, and 
transmutations matrix curves to determine input and output isotopic compositions. The reactor
can also operate with user-defined fuel compositions, called the "forward" mode. 

**Bright-lite Fuel Fabrication Facility**: A module that communicates directly with the Bright-lite
Reactor Facility that allows for the reactor facility to access blending functions that will 
determine the isotopic composition of fuel that the reactor will be using in order to match a
specific constraint. 

**Bright-lite Reprocessing Facility**: A module that seperates user-defined isotopes out of a material
that is sent to it. This currently requires an external text file to indicate which isotopes
should be seperated with what efficiency into which streams. The structure of this text file can be seen here_.

Bright-lite works in conjuction with the Cyclus_ Fuel Cycle Simulator. 

Bright-lite is currently only being actively supported for Ubuntu.

============
Installation
============
To use Bright-lite first you need to install it. Currently Bright-lite has
the following dependencies. 

- Cyclus_ Fuel Cycle Simulator

The following dependencies will be required in the future

- Eigen_

To install Bright-lite please follow these instructions.

1) Clone the Bright-lite repository from github. 

    git clone **repository url**

2) Change directory into the Bright-lite directory using the following
   command. 
    cd Bright-lite
   	
3) Use the following command inside the Bright-lite directory.
    python install.py
   	
This will add the Bright-lite module to the cyclus environment, and allow
you to use Bright-lite in Cyclus simulations. 



============
Using Bright-lite
============
Bright-lite reactor requires at least 6 inputs from the users to operate fully. While
there are several other inputs associated with the module, all of 
these other inputs come with a default value. 

-----------
The six required inputs are
-----------

- **in_commods**: This field is a one or more that indicates the possible sources of 
  fuel for the reactor. The values in this field should be commodities that exist 
  inside of the simulation. In order to use Bright-lite in forward mode, set the first
  source to the steady-state (non-startup) fuel supplier. For forward mode the startup
  fuels can be set with additional sources.
- **out_commod**: This field should be filled out with the cyclus commodity that will
  connect the reactor facility to the facility that will be directly handling the 
  waste.
- **libraries**: This is a one or more field that indicates the Bright-lite library 
  database the reactor will be using. Note: Adding additionally library databases to this list
  will enable the library interpolation_ capabilities in Bright-lite but also
  requires that the user input parameters and values to be interpolated upon. The
  interpolation feature is intended for advanced users. 
- **target_burnup**: This field indicates the target burnup for the 
  reactor. If this is set to 0, the reactor will operate in forward mode. If 
  this value is not set to zero the Bright-lite reactor must be connected to a
  Bright-lite fuel fabrication facility.
- **core_mass**: This field indicates the total mass of fuel inside of the core. This mass
  does not include structural components, it is only the mass of fuel to be burned.
- **generated_power**: This indicates the total thermal generating power of the core. 
  The electrical generated_power will be this value times the effiency of the reactor
  (a input set to default at 33% but is user adjustable).

-----------
Operational Modes
-----------
Bright-lite has two operational modes. The mode is indicated using inputs to the 
Bright-lite reactor module. Forward mode is chosen by setting the reactor *target_burnup*
mode to be equal to 0. Blending mode requires the *target_burnup* field to be a non negative
value. Additionally, blending mode requires the Bright-lite ReactorFacility to be connected to a
Bright-lite FuelfabFacility. 
 
^^^^^^^^^^^^
Forward Mode
^^^^^^^^^^^^
In forward mode Bright-lite accepts a fuel composition and burns it. 
It does this by advancing the fluence of each batch in the core until the target is met (such as k = 1).

Currently forward mode works only with *criticality* and *burnup* targets.

^^^^^^^^^^^^
Blending Mode
^^^^^^^^^^^^
As stated above using the blending function in Bright-lite requires connecting a Bright-lite
ReactorFacility to a Bright-lite FuelfabFacility. The *in_commods* field of the Bright-lite
reactor should include all of the fuel fabrication facilities that the reactor can be connected to. 

Currently there are two blending modes available in Bright-lite. These modes are described by
a target-constraint pairing. The two available pairs currently are:
 
1) Burnup - Criticality: The blender will create a fuel that meets a target burnup when criticality is equal to the given constraint. This set of constraints only requires a non negative number to be entered into the *target_burnup* field. 
2) Burnup - Conversion Ratio: The blender will create a fuel that meets a target burnup when conversion ratio is equal to the given constraint. This is achieved by setting the *CR_target* input field of the reactor to 
be equal to a number greater than 0 (note that there is no upper bound limit in the code for this this but physically it should not exceed 2). Additionally the *target_burnup* field must be a non negative value for this to work. 

------------
Running Example Cases
------------
There are several example cases provided with Bright-lite. The single reactor example cases are:

* LWR, 3 batch, 33 MWd/kg burnup, 3.1 % U-235 fuel (forward mode)
* LWR, 4 batch, 42 MWd/kg burnup, 3.6 % U-235 fuel (forward mode)
* LWR, 4 batch, 51 MWd/kg burnup, 4.3 % U-235 fuel (forward mode)
* LWR, 4 batch, 45 MWd/kg target burnup (blending mode)
* MOX, 5 batch, 50 MWd/kg burnup (forward mode)
* MOX, 5 batch, 50 MWd/kg target burnup (blending mode)
* Fast reactor, 6 batch, 180 MWd/kg burnup (forward mode)

The files can be found in the Bright-lite/examples/ folder. Run the following command to run the 42 MWd/kg burnup example case.

    cyclus Bright-lite/examples/LWR42forward.xml

------------
Library Interpolation
------------
.. _interpolation:

The libraries_ used in Bright-lite are often associated with several parameters. For example
an LWR reactor library might have parameters for burnup, and enrichment. If as a user, you

CHANGELOG
======================
Bright-lite Change Log
======================

.. current developments

v0.9
====================





nucname.cc
// Converts between naming conventions for nuclides.
// zzaaam is for numerals only (923350).
// name is for letters as well (U-235).
// MCNP is for numerals without the meta-stable flag (92235), as used in MCNP.

#include "nucname.h"

/*** Constructs the LL to zz Dictionary ***/
pyne::nucname::name_zz_t pyne::nucname::get_name_zz()
{
  pyne::nucname::name_zz_t lzd;

  lzd["Be"] = 04;
  lzd["Ba"] = 56;
  lzd["Bh"] = 107;
  lzd["Bi"] = 83;
  lzd["Bk"] = 97;
  lzd["Br"] = 35;
  lzd["Ru"] = 44;
  lzd["Re"] = 75;
  lzd["Rf"] = 104;
  lzd["Rg"] = 111;
  lzd["Ra"] = 88;
  lzd["Rb"] = 37;
  lzd["Rn"] = 86;
  lzd["Rh"] = 45;
  lzd["Tm"] = 69;
  lzd["H"] = 01;
  lzd["P"] = 15;
  lzd["Ge"] = 32;
  lzd["Gd"] = 64;
  lzd["Ga"] = 31;
  lzd["Os"] = 76;
  lzd["Hs"] = 108;
  lzd["Zn"] = 30;
  lzd["Ho"] = 67;
  lzd["Hf"] = 72;
  lzd["Hg"] = 80;
  lzd["He"] = 02;
  lzd["Pr"] = 59;
  lzd["Pt"] = 78;
  lzd["Pu"] = 94;
  lzd["Pb"] = 82;
  lzd["Pa"] = 91;
  lzd["Pd"] = 46;
  lzd["Po"] = 84;
  lzd["Pm"] = 61;
  lzd["C"] = 6;
  lzd["K"] = 19;
  lzd["O"] = 8;
  lzd["S"] = 16;
  lzd["W"] = 74;
  lzd["Eu"] = 63;
  lzd["Es"] = 99;
  lzd["Er"] = 68;
  lzd["Md"] = 101;
  lzd["Mg"] = 12;
  lzd["Mo"] = 42;
  lzd["Mn"] = 25;
  lzd["Mt"] = 109;
  lzd["U"] = 92;
  lzd["Fr"] = 87;
  lzd["Fe"] = 26;
  lzd["Fm"] = 100;
  lzd["Ni"] = 28;
  lzd["No"] = 102;
  lzd["Na"] = 11;
  lzd["Nb"] = 41;
  lzd["Nd"] = 60;
  lzd["Ne"] = 10;
  lzd["Zr"] = 40;
  lzd["Np"] = 93;
  lzd["B"] = 05;
  lzd["Co"] = 27;
  lzd["Cm"] = 96;
  lzd["F"] = 9;
  lzd["Ca"] = 20;
  lzd["Cf"] = 98;
  lzd["Ce"] = 58;
  lzd["Cd"] = 48;
  lzd["V"] = 23;
  lzd["Cs"] = 55;
  lzd["Cr"] = 24;
  lzd["Cu"] = 29;
  lzd["Sr"] = 38;
  lzd["Kr"] = 36;
  lzd["Si"] = 14;
  lzd["Sn"] = 50;
  lzd["Sm"] = 62;
  lzd["Sc"] = 21;
  lzd["Sb"] = 51;
  lzd["Sg"] = 106;
  lzd["Se"] = 34;
  lzd["Yb"] = 70;
  lzd["Db"] = 105;
  lzd["Dy"] = 66;
  lzd["Ds"] = 110;
  lzd["La"] = 57;
  lzd["Cl"] = 17;
  lzd["Li"] = 03;
  lzd["Tl"] = 81;
  lzd["Lu"] = 71;
  lzd["Lr"] = 103;
  lzd["Th"] = 90;
  lzd["Ti"] = 22;
  lzd["Te"] = 52;
  lzd["Tb"] = 65;
  lzd["Tc"] = 43;
  lzd["Ta"] = 73;
  lzd["Ac"] = 89;
  lzd["Ag"] = 47;
  lzd["I"] = 53;
  lzd["Ir"] = 77;
  lzd["Am"] = 95;
  lzd["Al"] = 13;
  lzd["As"] = 33;
  lzd["Ar"] = 18;
  lzd["Au"] = 79;
  lzd["At"] = 85;
  lzd["In"] = 49;
  lzd["Y"] = 39;
  lzd["N"] = 07;
  lzd["Xe"] = 54;
  lzd["Cn"] = 112;
  lzd["Fl"] = 114;
  lzd["Lv"] = 116;

  return lzd;
};
pyne::nucname::name_zz_t pyne::nucname::name_zz = pyne::nucname::get_name_zz();


/*** Constructs zz to LL dictionary **/
pyne::nucname::zzname_t pyne::nucname::get_zz_name()
{
  zzname_t zld;
  for (name_zz_iter i = name_zz.begin(); i != name_zz.end(); i++)
  {
    zld[i->second] = i->first;
  }
  return zld;
};
pyne::nucname::zzname_t pyne::nucname::zz_name = pyne::nucname::get_zz_name();



/******************************************/
/*** Define useful elemental group sets ***/
/******************************************/

pyne::nucname::zz_group pyne::nucname::name_to_zz_group(pyne::nucname::name_group eg)
{
  zz_group zg;
  for (name_group_iter i = eg.begin(); i != eg.end(); i++)
    zg.insert(name_zz[*i]);
  return zg;
};

// Lanthanides
pyne::nucname::name_t pyne::nucname::LAN_array[15] = {"La", "Ce", "Pr", "Nd",
  "Pm", "Sm", "Eu", "Gd", "Tb", "Dy", "Ho", "Er", "Tm", "Yb", "Lu"};
pyne::nucname::name_group pyne::nucname::LAN (pyne::nucname::LAN_array,
                                              pyne::nucname::LAN_array+15);
pyne::nucname::zz_group pyne::nucname::lan = \
  pyne::nucname::name_to_zz_group(pyne::nucname::LAN);

// Actinides
pyne::nucname::name_t pyne::nucname::ACT_array[15] = {"Ac", "Th", "Pa", "U",
  "Np", "Pu", "Am", "Cm", "Bk", "Cf", "Es", "Fm", "Md", "No", "Lr"};
pyne::nucname::name_group pyne::nucname::ACT (pyne::nucname::ACT_array, pyne::nucname::ACT_array+15);
pyne::nucname::zz_group pyne::nucname::act = pyne::nucname::name_to_zz_group(pyne::nucname::ACT);

// Transuarnics
pyne::nucname::name_t pyne::nucname::TRU_array[22] = {"Np", "Pu", "Am", "Cm",
  "Bk", "Cf", "Es", "Fm", "Md", "No", "Lr", "Rf", "Db", "Sg", "Bh", "Hs", "Mt",
  "Ds", "Rg", "Cn", "Fl", "Lv"};
pyne::nucname::name_group pyne::nucname::TRU (pyne::nucname::TRU_array,
                                              pyne::nucname::TRU_array+22);
pyne::nucname::zz_group pyne::nucname::tru = \
  pyne::nucname::name_to_zz_group(pyne::nucname::TRU);

//Minor Actinides
pyne::nucname::name_t pyne::nucname::MA_array[10] = {"Np", "Am", "Cm", "Bk",
  "Cf", "Es", "Fm", "Md", "No", "Lr"};
pyne::nucname::name_group pyne::nucname::MA (pyne::nucname::MA_array,
                                             pyne::nucname::MA_array+10);
pyne::nucname::zz_group pyne::nucname::ma = \
  pyne::nucname::name_to_zz_group(pyne::nucname::MA);

//Fission Products
pyne::nucname::name_t pyne::nucname::FP_array[88] = {"Ag", "Al", "Ar", "As",
  "At", "Au", "B", "Ba", "Be", "Bi", "Br", "C", "Ca", "Cd", "Ce", "Cl", "Co",
  "Cr", "Cs", "Cu", "Dy", "Er", "Eu", "F", "Fe", "Fr", "Ga", "Gd", "Ge", "H",
  "He", "Hf", "Hg", "Ho", "I", "In", "Ir", "K", "Kr", "La", "Li", "Lu", "Mg",
  "Mn", "Mo", "N", "Na", "Nb", "Nd", "Ne", "Ni", "O", "Os", "P", "Pb", "Pd",
  "Pm", "Po", "Pr", "Pt", "Ra", "Rb", "Re", "Rh", "Rn", "Ru", "S", "Sb", "Sc",
  "Se", "Si", "Sm", "Sn", "Sr", "Ta", "Tb", "Tc", "Te", "Ti", "Tl", "Tm", "V",
  "W", "Xe", "Y", "Yb", "Zn", "Zr"};
pyne::nucname::name_group pyne::nucname::FP (pyne::nucname::FP_array,
                                             pyne::nucname::FP_array+88);
pyne::nucname::zz_group pyne::nucname::fp = \
  pyne::nucname::name_to_zz_group(pyne::nucname::FP);


/***************************/
/*** isnuclide functions ***/
/***************************/

bool pyne::nucname::isnuclide(std::string nuc)
{
  int n;
  try {
    n = id(nuc);
  }
  catch(NotANuclide) {
    return false;
  };
  return isnuclide(n);
};

bool pyne::nucname::isnuclide(char * nuc)
{
  return isnuclide(std::string(nuc));
};

bool pyne::nucname::isnuclide(int nuc)
{
  int n;
  try {
    n = id(nuc);
  }
  catch(NotANuclide) {
    return false;
  };
  int zzz = n / 10000000;
  int aaa = (n % 10000000) / 10000;
  if (aaa == 0)
    return true;
  else if (aaa < zzz)
    return false;
  return true;
};



/********************/
/*** id functions ***/
/********************/
int pyne::nucname::id(int nuc)
{
  int newnuc;
  int zzz = nuc / 10000000; // ZZZ ?
  int aaassss = nuc % 10000000; // AAA-SSSS ?
  int aaa = aaassss / 10000; // AAA ?
  // Nuclide must already be in id form
  if (0 < zzz && zzz <= aaa && aaa <= zzz * 7)
  {
    // Normal nuclide
    return nuc;
  }
  else if (aaassss == 0 && 0 < zz_name.count(zzz))
  {
    // Natural elemental nuclide: ie for Urnaium = 920000000
    return nuc;
  }
  else if (nuc < 1000 && 0 < zz_name.count(nuc))
    // Gave Z-number
    return nuc * 10000000;

  // Not in id form, try ZZZAAAM form.
  zzz = nuc / 10000; // ZZZ ?
  aaassss = nuc % 10000; // AAA-SSSS ?
  aaa = aaassss / 10; // AAA ?
  if (zzz <= aaa && aaa <= zzz * 7)
  {
    // ZZZAAAM nuclide
    return (zzz*10000000) + (aaa*10000) + (nuc%10);
  }
  else if (aaa <= zzz && zzz <= aaa * 7 && 0 < zz_name.count(aaa))
  {
    // Cinder-form (aaazzzm), ie 2350920
    return (aaa*10000000) + (zzz*10000) + (nuc%10);
  }
  //else if (aaassss == 0 && 0 == zz_name.count(nuc/1000) && 0 < zz_name.count(zzz))
  else if (aaassss == 0 && 0 < zz_name.count(zzz))
  {
    // zzaaam form natural nuclide
    return zzz * 10000000;
  }

  if (nuc >= 1000000){
    // From now we assume no metastable info has been given.
    throw IndeterminateNuclideForm(nuc, "");
  };

  // Nuclide is not in zzaaam form,
  // Try MCNP form, ie zzaaa
  zzz = nuc / 1000;
  aaa = nuc % 1000;
  if (zzz <= aaa)
  {
    if (aaa - 400 < 0)
    {
      if (nuc == 95242)
        return nuc * 10000 + 1; // special case MCNP Am-242m
      else
        return nuc * 10000; // Nuclide in normal MCNP form
    }
    else
    {
      // Nuclide in MCNP metastable form
      if (nuc == 95642)
        return (95642 - 400)*10000; // special case MCNP Am-242
      nuc = ((nuc - 400) * 10000) + 1;
      while (3.0 < (float ((nuc/10000)%1000) / float (nuc/10000000)))
        nuc -= 999999;
      return nuc;
    }
  }
  else if (aaa == 0 && 0 < zz_name.count(zzz))
    // MCNP form natural nuclide
    return zzz * 10000000;

  // Not a normal nuclide, might be a
  // Natural elemental nuclide.
  // ie 92 for Urnaium = 920000
  if (0 < zz_name.count(nuc))
    return nuc * 10000000;
  throw IndeterminateNuclideForm(nuc, "");
};

int pyne::nucname::id(char * nuc)
{
  std::string newnuc (nuc);
  return id(newnuc);
};

int pyne::nucname::id(std::string nuc)
{
  if (nuc.empty())
    throw NotANuclide(nuc, "<empty>");
  int newnuc;
  std::string elem_name;

  // Get the string into a regular form
  std::string nucstr = pyne::to_upper(nuc);
  nucstr = pyne::remove_substring(nucstr, "-");
  int nuclen = nucstr.length();

  if (pyne::contains_substring(pyne::digits, nucstr.substr(0, 1)))
  {
    if (pyne::contains_substring(pyne::digits, nucstr.substr(nuclen-1, nuclen)))
    {
      // Nuclide must actually be an integer that
      // just happens to be living in string form.
      newnuc = pyne::to_int(nucstr);
      newnuc = id(newnuc);
    }
    else
    {
      // probably in NIST-like form (242Am)
      // Here we know we have both digits and letters
      std::string anum_str = pyne::remove_characters(nucstr, pyne::alphabet);
      newnuc = pyne::to_int(anum_str) * 10000;

      // Add the Z-number
      elem_name = pyne::remove_characters(nucstr, pyne::digits);
      elem_name = pyne::capitalize(elem_name);
      if (0 < name_zz.count(elem_name))
        newnuc = (10000000 * name_zz[elem_name]) + newnuc;
      else
        throw NotANuclide(nucstr, newnuc);
    };
  }
  else if (pyne::contains_substring(pyne::alphabet, nucstr.substr(0, 1)))
  {
    // Nuclide is probably in name form, or some variation therein
    std::string anum_str = pyne::remove_characters(nucstr, pyne::alphabet);

    // natural element form, a la 'U' -> 920000000
    if (anum_str.empty()) {
      elem_name = pyne::capitalize(nucstr);
      if (0 < name_zz.count(elem_name))
        return 10000000 * name_zz[elem_name];
    }

    int anum = pyne::to_int(anum_str);

    // bad form
    if (anum < 0)
      throw NotANuclide(nucstr, anum);

    // Figure out if we are meta-stable or not
    std::string end_char = pyne::last_char(nucstr);
    if (end_char == "M")
      newnuc = (10000 * anum) + 1;
    else if (pyne::contains_substring(pyne::digits, end_char))
      newnuc = (10000 * anum);
    else
      throw NotANuclide(nucstr, newnuc);

    // Add the Z-number
    elem_name = pyne::remove_characters(nucstr.substr(0, nuclen-1), pyne::digits);
    elem_name = pyne::capitalize(elem_name);
    if (0 < name_zz.count(elem_name))
      newnuc = (10000000 * name_zz[elem_name]) + newnuc;
    else
      throw NotANuclide(nucstr, newnuc);
  }
  else
  {
    // Clearly not a nuclide
    throw NotANuclide(nuc, nucstr);
  }
  return newnuc;
};



/**********************/
/*** name functions ***/
/**********************/
std::string pyne::nucname::name(int nuc)
{
  int nucid = id(nuc);
  std::string newnuc = "";

  int zzz = nucid / 10000000;
  int ssss = nucid % 10000;
  int aaassss = nucid % 10000000;
  int aaa = aaassss / 10000;

  // Make sure the LL value is correct
  if (0 == zz_name.count(zzz))
    throw NotANuclide(nuc, nucid);

  // Add LL
  newnuc += zz_name[zzz];

  // Add A-number
  if (0 < aaa)
    newnuc += pyne::to_str(aaa);

  // Add meta-stable flag
  if (0 < ssss)
    newnuc += "M";

  return newnuc;
};



std::string pyne::nucname::name(char * nuc)
{
  std::string newnuc (nuc);
  return name(newnuc);
}


std::string pyne::nucname::name(std::string nuc)
{
  return name(id(nuc));
}


/**********************/
/*** znum functions ***/
/**********************/
int pyne::nucname::znum(int nuc){
  return id(nuc) / 10000000;
};

int pyne::nucname::znum(char * nuc){
  return id(nuc) / 10000000;
};

int pyne::nucname::znum(std::string nuc){
  return id(nuc) / 10000000;
};

/**********************/
/*** anum functions ***/
/**********************/
int pyne::nucname::anum(int nuc){
  return (id(nuc) / 10000) % 1000;
};

int pyne::nucname::anum(char * nuc){
  return (id(nuc) / 10000) % 1000;
};

int pyne::nucname::anum(std::string nuc){
  return (id(nuc) / 10000) % 1000;
};

/**********************/
/*** snum functions ***/
/**********************/
int pyne::nucname::snum(int nuc){
  return id(nuc) % 10000;
};

int pyne::nucname::snum(char * nuc){
  return id(nuc) % 10000;
};

int pyne::nucname::snum(std::string nuc){
  return id(nuc) % 10000;
};

/************************/
/*** zzaaam functions ***/
/************************/
int pyne::nucname::zzaaam(int nuc)
{
  int nucid = id(nuc);
  int zzzaaa = nucid / 10000;
  int ssss = nucid % 10000;
  if (10 <= ssss)
    ssss = 9;
  return zzzaaa*10 + ssss;
};


int pyne::nucname::zzaaam(char * nuc)
{
  std::string newnuc (nuc);
  return zzaaam(newnuc);
};


int pyne::nucname::zzaaam(std::string nuc)
{
  return zzaaam(id(nuc));
};


int pyne::nucname::zzaaam_to_id(int nuc)
{
  return (nuc/10)*10000 + (nuc%10);
};


int pyne::nucname::zzaaam_to_id(char * nuc)
{
  return zzaaam_to_id(std::string(nuc));
};


int pyne::nucname::zzaaam_to_id(std::string nuc)
{
  return zzaaam_to_id(pyne::to_int(nuc));
};


/**********************/
/*** mcnp functions ***/
/**********************/
int pyne::nucname::mcnp(int nuc)
{
  nuc = id(nuc);
  int ssss = nuc % 10000;
  int newnuc = nuc / 10000;

  // special case Am242(m)
  if (newnuc == 95242 && ssss < 2)
    ssss = (ssss + 1) % 2;

  // Handle the crazy MCNP meta-stable format
  if (0 != ssss && ssss < 10)
    newnuc += 300 + (ssss * 100);

  return newnuc;
};



int pyne::nucname::mcnp(char * nuc)
{
  std::string newnuc (nuc);
  return mcnp(newnuc);
};



int pyne::nucname::mcnp(std::string nuc)
{
  return mcnp(id(nuc));
};

//
// MCNP -> id
//
int pyne::nucname::mcnp_to_id(int nuc)
{
  int zzz = nuc / 1000;
  int aaa = nuc % 1000;
  if (zzz <= aaa)
  {
    if (aaa - 400 < 0)
    {
      if (nuc == 95242)
        return nuc * 10000 + 1; // special case MCNP Am-242m
      else
        return nuc * 10000; // Nuclide in normal MCNP form
    }
    else
    {
      // Nuclide in MCNP metastable form
      if (nuc == 95642)
        return (95642 - 400)*10000; // special case MCNP Am-242
      nuc = ((nuc - 400) * 10000) + 1;
      while (3.0 < (float ((nuc/10000)%1000) / float (nuc/10000000)))
        nuc -= 999999;
      return nuc;
    }
  }
  else if (aaa == 0)
    // MCNP form natural nuclide
    return zzz * 10000000;
};


int pyne::nucname::mcnp_to_id(char * nuc)
{
  return mcnp_to_id(std::string(nuc));
};


int pyne::nucname::mcnp_to_id(std::string nuc)
{
  return mcnp_to_id(pyne::to_int(nuc));
};


/*************************/
/*** serpent functions ***/
/*************************/
std::string pyne::nucname::serpent(int nuc)
{
  int nucid = id(nuc);
  std::string newnuc = "";

  int ssss = nucid % 10000;
  int aaassss = nucid % 10000000;
  int zzz = nucid / 10000000;
  int aaa = aaassss / 10000;

  // Make sure the LL value is correct
  if (0 == zz_name.count(zzz))
    throw NotANuclide(nuc, nucid);

  // Add LL
  std::string llupper = pyne::to_upper(zz_name[zzz]);
  std::string lllower = pyne::to_lower(zz_name[zzz]);
  newnuc += llupper[0];
  for (int l = 1; l < lllower.size(); l++)
    newnuc += lllower[l];

  // Add required dash
  newnuc += "-";

  // Add A-number
  if (0 < aaassss)
    newnuc += pyne::to_str(aaa);
  else if (0 == aaassss)
    newnuc += "nat";

  // Add meta-stable flag
  if (0 < ssss)
    newnuc += "m";

  return newnuc;
};


std::string pyne::nucname::serpent(char * nuc)
{
  std::string newnuc (nuc);
  return serpent(newnuc);
};


std::string pyne::nucname::serpent(std::string nuc)
{
  return serpent(id(nuc));
};

//
// Serpent -> id
//
//int pyne::nucname::serpent_to_id(int nuc)
//{
// Should be ZAID
//};


int pyne::nucname::serpent_to_id(char * nuc)
{
  return serpent_to_id(std::string(nuc));
};


int pyne::nucname::serpent_to_id(std::string nuc)
{
  if (nuc.empty())
    throw NotANuclide(nuc, "<empty>");
  int nucid;
  std::string elem_name;

  // Get the string into a regular form
  std::string nucstr = pyne::to_upper(nuc);
  nucstr = pyne::remove_substring(nucstr, "-");
  int nuclen = nucstr.length();

  // Nuclide is probably in name form, or some variation therein
  std::string anum_str = pyne::remove_characters(nucstr, pyne::alphabet);

  // natural element form, a la 'U' -> 920000000
  if (anum_str.empty() || pyne::contains_substring(nucstr, "NAT")) {
    elem_name = pyne::capitalize(pyne::remove_substring(nucstr, "NAT"));
    if (0 < name_zz.count(elem_name))
      return 10000000 * name_zz[elem_name];
  }
  int anum = pyne::to_int(anum_str);

  // Figure out if we are meta-stable or not
  std::string end_char = pyne::last_char(nucstr);
  if (end_char == "M")
    nucid = (10000 * anum) + 1;
  else if (pyne::contains_substring(pyne::digits, end_char))
    nucid = (10000 * anum);
  else
    throw NotANuclide(nucstr, nucid);

  // Add the Z-number
  elem_name = pyne::remove_characters(nucstr.substr(0, nuclen-1), pyne::digits);
  elem_name = pyne::capitalize(elem_name);
  if (0 < name_zz.count(elem_name))
    nucid = (10000000 * name_zz[elem_name]) + nucid;
  else
    throw NotANuclide(nucstr, nucid);
  return nucid;
};


/**********************/
/*** nist functions ***/
/**********************/
std::string pyne::nucname::nist(int nuc)
{
  int nucid = id(nuc);
  std::string newnuc = "";

  int zzz = nucid / 10000000;
  int ssss = nucid % 10000;
  int aaassss = nucid % 10000000;
  int aaa = aaassss / 10000;

  // Make sure the LL value is correct
  if (0 == zz_name.count(zzz))
    throw NotANuclide(nuc, nucid);

  // Add A-number
  if (0 < aaassss)
    newnuc += pyne::to_str(aaa);

  // Add name
  std::string name_upper = pyne::to_upper(zz_name[zzz]);
  std::string name_lower = pyne::to_lower(zz_name[zzz]);
  newnuc += name_upper[0];
  for (int l = 1; l < name_lower.size(); l++)
    newnuc += name_lower[l];

  // Add meta-stable flag
  // No metastable flag for NIST,
  // but could add star, by uncommenting below
  //if (0 < mod_10)
  // newnuc += "*";

  return newnuc;
};


std::string pyne::nucname::nist(char * nuc)
{
  std::string newnuc (nuc);
  return nist(newnuc);
};


std::string pyne::nucname::nist(std::string nuc)
{
  return nist(id(nuc));
};


//
// NIST -> id
//
//int pyne::nucname::nist_to_id(int nuc)
//{
// NON-EXISTANT
//};

int pyne::nucname::nist_to_id(char * nuc)
{
  return nist_to_id(std::string(nuc));
};

int pyne::nucname::nist_to_id(std::string nuc)
{
  if (nuc.empty())
    throw NotANuclide(nuc, "<empty>");
  int nucid;
  nuc = pyne::to_upper(nuc);
  std::string elem_name;
  int nuclen = nuc.length();

  // Nuclide is probably in name form, or some variation therein
  std::string anum_str = pyne::remove_characters(nuc, pyne::alphabet);

  // natural element form, a la 'U' -> 920000000
  if (anum_str.empty()) {
    elem_name = pyne::capitalize(nuc);
    if (0 < name_zz.count(elem_name))
      return 10000000 * name_zz[elem_name];
  }
  nucid = pyne::to_int(anum_str) * 10000;

  // Add the Z-number
  elem_name = pyne::remove_characters(nuc, pyne::digits);
  elem_name = pyne::capitalize(elem_name);
  if (0 < name_zz.count(elem_name))
    nucid = (10000000 * name_zz[elem_name]) + nucid;
  else
    throw NotANuclide(nuc, nucid);
  return nucid;
};


/************************/
/*** cinder functions ***/
/************************/
int pyne::nucname::cinder(int nuc)
{
  // cinder nuclides of form aaazzzm
  int nucid = id(nuc);
  int zzz = nucid / 10000000;
  int ssss = nucid % 10000;
  int aaassss = nucid % 10000000;
  int aaa = aaassss / 10000;
  if (10 <= ssss)
    ssss = 9;
  return (aaa*10000) + (zzz*10) + ssss;
};



int pyne::nucname::cinder(char * nuc)
{
  std::string newnuc (nuc);
  return cinder(newnuc);
};



int pyne::nucname::cinder(std::string nuc)
{
  return cinder(id(nuc));
};

//
// Cinder -> Id
//
int pyne::nucname::cinder_to_id(int nuc)
{
  int ssss = nuc % 10;
  int aaazzz = nuc / 10;
  int zzz = aaazzz % 1000;
  int aaa = aaazzz / 1000;
  return (zzz * 10000000) + (aaa * 10000) + ssss;
};


int pyne::nucname::cinder_to_id(char * nuc)
{
  return cinder_to_id(std::string(nuc));
};


int pyne::nucname::cinder_to_id(std::string nuc)
{
  return cinder_to_id(pyne::to_int(nuc));
};




/**********************/
/*** ALARA functions ***/
/**********************/
std::string pyne::nucname::alara(int nuc)
{
  int nucid = id(nuc);
  std::string newnuc = "";
  std::string ll = "";

  int zzz = nucid / 10000000;
  int ssss = nucid % 10000;
  int aaassss = nucid % 10000000;
  int aaa = aaassss / 10000;

  // Make sure the LL value is correct
  if (0 == zz_name.count(zzz))
    throw NotANuclide(nuc, nucid);

  // Add LL, in lower case
  ll += zz_name[zzz];

  for(int i = 0; ll[i] != '\0'; i++)
    ll[i] = tolower(ll[i]);
  newnuc += ll;

  // Add A-number
  if (0 < aaassss){
    newnuc += ":";
    newnuc += pyne::to_str(aaa);
  }

  // Note, ALARA input format does not use metastable flag
  return newnuc;
};


std::string pyne::nucname::alara(char * nuc)
{
  std::string newnuc (nuc);
  return alara(newnuc);
}


std::string pyne::nucname::alara(std::string nuc)
{
  return alara(id(nuc));
}


//
// Cinder -> Id
//
//int pyne::nucname::alara_to_id(int nuc)
//{
// Not Possible
//};


int pyne::nucname::alara_to_id(char * nuc)
{
  return alara_to_id(std::string(nuc));
};


int pyne::nucname::alara_to_id(std::string nuc)
{
  if (nuc.empty())
    throw NotANuclide(nuc, "<empty>");
  int nucid;
  nuc = pyne::to_upper(pyne::remove_characters(nuc, ":"));
  std::string elem_name;
  int nuclen = nuc.length();

  // Nuclide is probably in name form, or some variation therein
  std::string anum_str = pyne::remove_characters(nuc, pyne::alphabet);

  // natural element form, a la 'U' -> 920000000
  if (anum_str.empty()) {
    elem_name = pyne::capitalize(nuc);
    if (0 < name_zz.count(elem_name))
      return 10000000 * name_zz[elem_name];
  }
  nucid = pyne::to_int(anum_str) * 10000;

  // Add the Z-number
  elem_name = pyne::remove_characters(nuc, pyne::digits);
  elem_name = pyne::capitalize(elem_name);
  if (0 < name_zz.count(elem_name))
    nucid = (10000000 * name_zz[elem_name]) + nucid;
  else
    throw NotANuclide(nuc, nucid);
  return nucid;
};


